Even faster C.Run by using no reflect at all in common cases

dolmen opened the pull request:
In `C.Run` bypass check of the signature of the `Run` method of `c.TB` for the common cases ([`testing.T.Run`](https://pkg.go.dev/testing/#T.Run), [`testing.B.Run`](https://pkg.go.dev/testing/#B.Run), [`quicktest.C.Run`](https://pkg.go.dev/github.com/frankban/quicktest#C.Run)) by using a typecast.

Note: this is an alternative version of #160 and replaces it. The first 2 commits are common.

This serie of patches shows the developement process:
* refactor C.Run to extract `getRunFuncSignature`
* add unit test for `getRunFuncSignature`
* refactor C.Run to add shortcuts for common cases

dolmen commented:
I have just added a [commit](https://github.com/frankban/quicktest/pull/165/commits/4b471b12553a0824aa862a1c6c20a8fb22b037ea) that factorizes the `qt.C` setup for both the `reflect` version and the shortcuts ones. This removes the duplicated code that #166 is partially factorizing using generics.

rogpeppe commented on quicktest_test.go:
```diff
@@ -527,6 +528,26 @@ want:
+		(*customT)(nil),   // a custom form of testing.TB with a non-standard Run method
+	} {
+		tbt := reflect.TypeOf(tb)
+		farg, err := qt.GetRunFuncSignature(tbt)
```
This test isn't great IMHO. Firstly, it tests GetRunFuncSignature which is an internal implementation detail (this module in general tries to test the public API as much as possible to allow later internal refactoring without changing tests). Secondly, GetRunFuncSignature won't even be invoked in most of the above cases, so I'm not sure why this is explicitly testing them.

dolmen replied:
You're right. I had written that test in the context of #160, but it isn't relevant anymore here.

Also, the extended tests and benchmarks you wrote in #166 would be welcome on a separate MR (written without generics) and I would be glad to rebase.

dolmen replied:
I have now removed the commit that added that test.

dolmen commented:
I've cleaned the PR:
* removed the test `TestGetRunFuncSig` as suggested by @rogpeppe
* removed the wrong shortcut for `Run(string, *qt.C)` to satify the new test in #167 (which should be merged first)
